### PR TITLE
ipc4: add mixin/mixout channel remapping support

### PR DESCRIPTION
**Pull request by serhiy-katsyuba-intel**
Adds channel remapping feature support to mixin/mixout.

Signed-off-by: Serhiy Katsyuba <serhiy.katsyuba@intel.com>

### CODE REVIEW

**Review by lgirdwood — COMMENTED** [review]
@lyakh can you check the sparse usage.
@mengdonglin I guess this will unblock you ?

**Comment by lgirdwood** [comment]
@wszypelt @lrudyX can someone check CI, not expecting this PR to fail at early test stage. Thanks !

**Comment by mengdonglin** [comment]
> @lyakh can you check the sparse usage. @mengdonglin I guess this will unblock you ?
@lgirdwood We're not blocked by this atm.

File: src/audio/mixin_mixout.c
Comment (lyakh) [inline src/audio/mixin_mixout.c:713]: I'm not a native English speaker but "normally this should never happen" makes me slightly nervous and makes me think about parallel universes and alternative realities :-)

File: src/audio/mixin_mixout.c
Comment (lyakh) [inline src/audio/mixin_mixout.c:1143]: you have several cases like
```
if (a) {
	do_stuff();
} else {
	print("error");
	return -EBAD;
}
```
whereas it's easier to write
```
if (!a) {
	print("error");
	return -EBAD;
}
do_stuff();
```

File: src/audio/mixin_mixout.c
Comment (lyakh) [inline src/audio/mixin_mixout.c:1211]: is it now going to be a different buffer? You probably cannot call `comp_verify_params()` while holding a buffer, because it tries to acquire buffers too? But at least you probably can reuse `sink`

File: src/audio/mixin_mixout.c
Comment (lyakh) [inline src/audio/mixin_mixout.c:1200]: is this needed because `comp_verify_params()` needs `sink_c->stream.frame_fmt`?

File: src/audio/mixin_mixout.c
Comment (lyakh) [inline src/audio/mixin_mixout.c:1440]: same comment about `if (!x) return -EBAD;`

**Review by lyakh — CHANGES_REQUESTED** [review]
I didn't check the actual logic, only buffer manipulations

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:1422]: Done.

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c]: Modified to reuse sink.

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c]: Done.

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:710]: Not sure what to do about this. Hope leaving as is is OK. Google knows 13600 occurrences of "normally this should never happen" with 164 of then on GitHub. Let it be one more :)

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:685]: Fixed.

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c]: Fixed.

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:606]: These are just renames, asserts() were here before this PR. And these asserts() are for conditions that can only happen due to developer mistake as a result of code modifications, not runtime configuration error.

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:285]: Answered few comments below.

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:293]: The loop could be simplified by using something like audio_stream_get_avail_frames(), audio_stream_get_free_frames(). But these two use division internally which is about 40 times slower then just pointer incrementation. So I chose to track pointers not reaching beyond end of buffer instead of calculating frames and tracking processed frames.
No, infinite lope is not the case here.

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:1194]: No, comp_verify_params() setups some buffer parameters (like frame_fmt), however, it does not setup valid_sample_fmt. This might be a bug in comp_verify_params(). That is why audio_stream_fmt_conversion() is called to properly set valid_sample_fmt. It also sets frame_fmt, but that is not really necessary as comp_verify_params() will do it once again. To summarize that audio_stream_fmt_conversion() call is a workaround to a missing functionality (a BUG maybe) in comp_verify_params().

File: src/audio/mixin_mixout.c
Comment (lyakh) [reply src/audio/mixin_mixout.c:293]: You're probably right, that the loop will never be infinite in practice, but it still looks risky to me. What it's doing there is something like
```
N=5;
while (N) {
	x=get_x();
	while (N && x) {
		N--;
	}
}
```
so, if `x` turns false before `N` is decremented to `0` the external loop will run forever. If we're absolutely sure, that the internal loop will be entered at least one for each external iteration, then we could at least replace the internal one with `do {} while (N && x)`

File: src/audio/mixin_mixout.c
Comment (lyakh) [reply src/audio/mixin_mixout.c:1194]: but then, if `comp_verify_params()` doesn't need it, can this be moved below to only acquire the `sink` buffer once?

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:293]: The problem with replacing while() {...} with do {...} while() is that then do {...} while() version will look like it is possible to write to a buffer (to *dst) without a prior check if pointer is within buffer boundaries. Consider a buffer size == 0 (which should never happen, of course), in this case while() {} will turn into infinite loop and do {} while () will write beyond buffer boundary. Writing beyond buffer is much worse (harder to debug) then infinite loop. So do {...} while () version is not better then while () {...}.

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:1200]: Done.

File: src/audio/mixin_mixout.c
Comment (lgirdwood) [inline src/audio/mixin_mixout.c:450]: Good idea to reuse, but I would rename the variables that are used only for remapping with a ```remap_``` prefix and make it obvious in the comment that remapping is enabled by setting them to non zero values.

Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:450]: @lgirdwood, I do not understand your suggestion. In channel remapping mode channels are mixed using mix_channel_sXX(); in normal mode, multichannel stream is treated as a single channel and so an entire stream is mixed using the same mix_channel_sXX() function dedicated to channel mixing. That function is unaware if it works on real channel of a multichannel stream or on a fake channel (a multichannel stream represented as single channel).
What variables do you suggest to mark with remap_ prefix?

Comment (lgirdwood) [reply src/audio/mixin_mixout.c:450]: The part that confuses me is the work "fake". I can also see 2 modes, NORMAL and REMAP (where you hard code 0, 1 in one of the modes). I would like to see it clear in the comment if you use mode X then you must set x,y = 0,1. This way everyone will understand how to use it.

File: src/audio/mixin_mixout.c
Comment (lyakh) [reply src/audio/mixin_mixout.c:293]: but you're saying that will never happen! :-) So, that's exactly my point. If there's any chance of an infinite loop there (which is the same as the problem you describe with my `do {} while ()` approach) - let's explicitly fix it now. If there's absolutely 0 such chance - then my `do {} while ()` isn't any worse and I'd say is better because it shows explicitly, that "we know what we're doing"

Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:293]: (1) Zero buffer length should not happen, as if there is no space in sink buffer mix_channel_sXX() will not be called at all.

(2) For zero buffer length case (which should never happen), while() {} will result in infinite loop, but do {} while() will result in writing to memory beyond buffer region. Both are problems, but do {} while() is worse (harder to notice, could result in strange "magic" bugs). So it is none benefits in replacing while () {} with do {} while().
I understand, looking into nested while() {} loops pair triggers a suspicious is there an infinite loop possible. However, looking into do {} while() with memory write inside triggers the suspicion has the pointer been checked to be within buffer range. 

(3) The function x=get_x() from your example above is x = audio_stream_wrap(..., x) and the job of that function is to make sure x condition in while() on next line is true for first iteration.

File: src/audio/mixin_mixout.c
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:453]: Added comment in mix_and_remap() how an entire stream is mixed reusing mix_channel() function.

**Comment by lgirdwood** [comment]
@lyakh all good now for you ?

**Review by lgirdwood — APPROVED** [review]

File: src/audio/mixin_mixout.c
Comment (lyakh) [inline src/audio/mixin_mixout.c:562]: `switch (sink_config->mixer_mode)`

**Comment by lyakh** [comment]
> @lyakh all good now for you ?

@lgirdwood looks better now, yes. A couple of points I'd try to improve a bit further, but nothing critical

**Pull request merged**
